wandb/sdk/lib/printer.py

```diff
@@ -0,0 +1,194 @@
+        off: Optional[bool] = None,
+    ) -> None:
+        if not off:
+            self._display(text, status=status)
+
+    @abstractmethod
+    def _display(
+        self, text: Union[str, List[str], Tuple[str]], *, status: Optional[str] = None,
+    ) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def code(self, text: str) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def name(self, text: str) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def link(self, link: str, text: Optional[str] = None) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def emoji(self, name: str) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def status(self, text: str, failure: Optional[bool] = None) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def files(self, text: str) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def grid(self, rows: List[List[str]], title: Optional[str] = None) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def panel(self, columns: List[str]) -> str:
+        raise NotImplementedError
+
+
+class PrinterTerm(_Printer):
+    def __init__(self) -> None:
+        super().__init__()
+        self._html = False
+        self._progress = itertools.cycle(["-", "\\", "|", "/"])
+
+    def _display(
+        self, text: Union[str, List[str], Tuple[str]], *, status: Optional[str] = None
+    ) -> None:
+        text = "\n".join(text) if isinstance(text, (list, tuple)) else text
+        if status == "info" or status is None:
+            wandb.termlog(text)
+        elif status == "warn":
+            wandb.termwarn(text)
+        elif status == "error":
+            wandb.termerror(text)
+        else:
+            raise
+
+    def progress_update(self, text: str, precentage: Optional[float] = None) -> None:
+        wandb.termlog(f"{next(self._progress)} {text}", newline=False)
+
+    def progress_close(self) -> None:
+        wandb.termlog(" " * 79)
+
+    def code(self, text: str) -> str:
+        return click.style(text, bold=True)
+
+    def name(self, text: str) -> str:
+        return click.style(text, fg="yellow")
+
+    def link(self, link: str, text: Optional[str] = None) -> str:
+        return click.style(link, fg="blue", underline=True)
+
+    def emoji(self, name: str) -> str:
+        emojis = dict()
+        if platform.system() != "Windows" and wandb.util.is_unicode_safe(sys.stdout):
+            emojis = dict(star="⭐️", broom="🧹", rocket="🚀")
+
+        return emojis.get(name, "")
+
+    def status(self, text: str, failure: Optional[bool] = None) -> str:
+        color = "red" if failure else "green"
+        return click.style(text, fg=color)
+
+    def files(self, text: str) -> str:
+        return click.style(text, fg="magenta", bold=True)
+
+    def grid(self, rows: List[List[str]], title: Optional[str] = None) -> str:
```

Comment:
soo sick!

Reply:
with rich it is so much prettier:
![image](https://user-images.githubusercontent.com/87335417/153443556-6ebb9a72-53dc-416e-b028-cc385c0324df.png)